odoo/addons/base/models/ir_mail_server.py

```diff
@@ -143,6 +143,46 @@ def _check_smtp_ssl_files(self):
             elif mail_server.smtp_ssl_private_key and not mail_server.smtp_ssl_certificate:
                 raise UserError(_('SSL certificate is missing for %s.', mail_server.name))
 
+    def write(self, vals):
+        """Ensure we cannot archive a server in-use"""
+        has_archive_error = False
+        if not vals.get('active', True):
+            usages = self._active_usages_compute()
+            if usages:
+                has_archive_error = True
+
+        if not has_archive_error:
+            return super().write(vals)
+
+        usages = {s: u for s, u in usages.items() if u}
+        usage_servers = []
+        usage_details = []
+        is_multiple_server_usage = len(usages) > 1
+        for record in sorted(self, key=lambda r: r.display_name):
+            if record.id not in usages:
+                continue
+            usage_servers.append(record.display_name)
```

Comment:
Can't you have usages keys being records, and then loop on the dictionary (having key = record, value = usages) ? That way we can get rid of usage_servers / usage_details and use the dictionary keys / values (manipulating a single variable is easier to understand and update the code). What do you think ?